kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,426 @@
+//! Defines [`LogReplayScanner`] used by [`TableChangesScan`] to process commit files and extract
+//! the metadata needed to generate the Change Data Feed.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, LazyLock};
+
+use crate::actions::visitors::{visit_deletion_vector_at, ProtocolVisitor};
+use crate::actions::{
+    get_log_add_schema, get_log_schema, Protocol, ADD_NAME, CDC_NAME, COMMIT_INFO_NAME,
+    METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME,
+};
+use crate::engine_data::{GetData, TypedGetData};
+use crate::expressions::{column_expr, column_name, ColumnName, Expression};
+use crate::path::ParsedLogPath;
+use crate::scan::data_skipping::DataSkippingFilter;
+use crate::scan::scan_row_schema;
+use crate::scan::state::DvInfo;
+use crate::schema::{ArrayType, ColumnNamesAndTypes, DataType, MapType, SchemaRef, StructType};
+use crate::table_properties::TableProperties;
+use crate::utils::require;
+use crate::{DeltaResult, Engine, EngineData, Error, ExpressionRef, RowVisitor};
+use itertools::Itertools;
+
+#[cfg(test)]
+mod tests;
+
+pub struct TableChangesScanData {
+    pub data: Box<dyn EngineData>,
+    // The selection vector used to filter log actions.
+    pub selection_vector: Vec<bool>,
+    // An optional map from a remove action's path to its deletion vector
+    pub remove_dv: Option<Arc<HashMap<String, DvInfo>>>,
+}
+
+/// Given an iterator of ParsedLogPath and a predicate, returns an iterator of
+/// [`TableChangesScanData`]  Each row that is selected in the returned
+/// `engine_data` _must_ be processed to complete the scan. Non-selected rows
+/// _must_ be ignored.
+pub(crate) fn table_changes_action_iter(
+    engine: Arc<dyn Engine>,
+    commit_files: impl IntoIterator<Item = ParsedLogPath>,
```

Comment:
commit files is iterated over once, but yeah I can change it to `Iterator`

Reply:
Actually, `IntoIterator` can only produce an iterator once, because `into_iter` consumes `self`. So it's purely for convenience of the caller (who can pass a reference to any container that shouldn't be consumed)

Reply:
yea sg let's revert to `IntoIterator` my bad!